### ISSUE TITLE
provide custom JSON stringifier

### ISSUE THREAD

**Issue by benjamine** [b123400/highland-json#24]
I see a few places are using `JSON.stringify`:

https://github.com/b123400/highland-json/blob/master/src/index.js#L24
https://github.com/b123400/highland-json/blob/master/src/index.js#L54
https://github.com/b123400/highland-json/blob/master/src/index.js#L62

it'd be useful to me (and suppose other users) to have an optional hook to do custom json serialization, 
In my case I'm interested in using https://github.com/fastify/fast-json-stringify here, mostly to still use a JSON schema for trimming output as part of the process.

An example API:
``` js
const fastAddressStringifier = fastStringify(addressSchema);

...
stringifyObj({
  // optional, receives the object to seralize, and the path that object is from the root
  stringify: (obj, path) => {
    if (path === 'users.address') {
       return fastAddressStringifier(obj);
    }
    return JSON.stringify(obj);
  },
}),
...

```

Or maybe a better alternative:
``` js
const fastAddressStringifier = fastStringify(addressSchema);

...
stringifyObj({
  // optional, receives the object to seralize, and the path that object is from the root
  stringifierByPath: path => {
    if (path === 'users.address') {
       return fastAddressStringifier;
    }
    return JSON.stringify;
  },
}),
...

```
(would you be open to a PR adding this capability?)

**Comment by b123400** [comment]
Such functionality would be nice! I think it's fine as long as it doesn't break backward compatibility. 

For example:

```javascript
_([1,2,3]).through(stringify) // use JSON.stringify

const options = {
  stringify: fastAddressStringifier
}
_([1,2,3]).through(stringify(options)) // use custom stringify function
```

However the "by path" idea is difficult to implement and I am against it.
Highland-json only "traverse" into some paths, but not all. It's difficult to guarantee `stringifierByPath` will be called with every path.

The idea of stream encoder is that, encoding can be started before all the data is received. If we already have the entire object, it makes sense to use a non-stream encoder and therefore out of highland-json's scope.

**Comment by benjamine** [comment]
thanks 👍 
so yes, I understand, I guess what I meant with that "path" arg is only to have a hint at where you are to know which serialization apply, an example:

``` json
{
  users: [ {...}, {...}, {...}, {...} ],
}
```
let's suppose this my document, and `users` is loaded as a highland stream, and I have a fastUserStringifier.
what I'd want is a way to know my stringifier is being invoked at the "users" path to pick the right stringifier, this doesn't mean it has to be called for every path, only at those where a stream exists.

In the case above there's only 1 stream (which tbf it's probably the most common right?), so no uncertainty there.
But, if there were multiple places where a stream exists, you can see how knowing in which part of the tree you are could be helpful to pick a custom stringifier.

maybe if we could send a reference to the source stream as second argument?

``` js
const doc = { numbers: _([1,2,3]), names: _(["john", "jill"]) };

const options = {
  stringify: (value, sourceStream) => sourceStream === doc.numbers ? fastNumberStringifier : fastNameStringifier
};

_([doc]).through(stringifyObj(options));
```

**Comment by b123400** [comment]
That sounds like a very good idea! 👍🏻

**Comment by b123400** [comment]
Implemented in 1.4.0
